test/jdk/java/util/Currency/ValidateISO4217.java

```diff
@@ -113,10 +140,86 @@ public class ValidateISO4217 {
             {"XK", "EUR", "978", "2"},      // Kosovo
     };
     private static SimpleDateFormat format = null;
+    private static final String MODULE_PATCH_LOCATION =
+            System.getProperty("test.classes") + "/java/util/";
+    private static final String MOCKED_TIME = System.getProperty("MOCKED.TIME");
+
+    // Classes that should mock System.currentTimeMillis()
+    private static final String[] CLASSES =
+            Stream.concat(
+                Stream.of("Currency.class"),
+                Arrays.stream(Currency.class.getDeclaredClasses())
+                        .map(c -> "Currency$" + c.getSimpleName() + ".class")
+            ).toArray(String[]::new);
+
+    // "check" invocation only runs the main method (and not any tests) to determine if the
+    // future time checking is correct
+    public static void main(String[] args) {
+        // Override for PK=PKR in test/currency.properties is PKZ - simple
+        // Override for PW=USD in test/currency.properties is MWP - special
+        if (MOCKED_TIME.equals("check")) {
+            assertEquals("PKZ", Currency.getInstance(Locale.of("", "PK")).getCurrencyCode(),
+                    "Mocked time / module patch not working");
+            assertEquals("MWP", Currency.getInstance(Locale.of("", "PW")).getCurrencyCode(),
+                    "Mocked time / module patch not working");
+            // Properly working. Do nothing and move to third invocation
+        } else {
+            throw new RuntimeException(
+                    "Incorrect usage of ValidateISO4217. Main method invoked without proper system property value");
+        }
+    }
+
+    @BeforeAll
+    static void setUp() throws Exception {
+        checkUsage();
+        setUpPatchedClasses();
+        setUpTestingData();
+    }
+
+    // Enforce correct usage of ValidateISO4217
+    static void checkUsage() {
+        if (MOCKED_TIME == null
+                || (!MOCKED_TIME.equals("setup") && !MOCKED_TIME.equals("true"))) {
+            throw new RuntimeException(
+                    "Incorrect usage of ValidateISO4217. Missing \"MOCKED.TIME\" system property");
+        }
+    }
+
+    // Patch the relevant classes required for module patch
+    static void setUpPatchedClasses() throws IOException {
+        if (MOCKED_TIME.equals("setup")) {
+            new File(MODULE_PATCH_LOCATION).mkdirs();
+            for (String s : CLASSES) {
+                patchClass(s);
+            }
+        }
+    }
+
+    // Mock calls of System.currentTimeMillis() within Currency to Long.MAX_VALUE.
+    // This effectively ensures that we are always past any cut-over dates.
+    private static void patchClass(String name) throws IOException {
+        CodeTransform codeTransform = (codeBuilder, e) -> {
+            switch (e) {
+                case InvokeInstruction i when i.name().stringValue().equals("currentTimeMillis") ->
```

Comment:
`equalsString()` may be used. Regardless, is there a way to tell this call is indeed `System.currentTimeMillis()`? Might do that check in case a method on Currency with the same name is introduced (not likely though).

Reply:
Yes, we can check the owner name is _java/lang/System_. Addressed your other comment as well in https://github.com/openjdk/jdk/pull/24701/commits/c58f11f2affa19267aa2416bcff10f842c13d871.